new exchange - bitget

yili1992 opened the pull request:
## Summary
to support trading for bitget 
## Quick changelog
- exchange/bitget
- tests

## What's new?
new exchange - bitget

Commit:
feat: add bitget

Commit:
feat: add bitget

Commit:
feat: add bitget

xmatthias commented on freqtrade/configuration/deploy_config.py:
```diff
@@ -111,6 +111,7 @@ def ask_user_config() -> dict[str, Any]:
                 "binance",
                 "binanceus",
                 "bingx",
+                "bitget",
```
we're only adding supported exchanges here - but i don't see bitget as becoming an officially supported exchange at this point (essentially that assumes that core team members have an account on that exchange and are able to actively support people with it).

Users can however still pick that exchange by selecting "other" and inserting that exchange's name.

yili1992 replied:
get it, i will remove it

xmatthias commented on freqtrade/exchange/bitget.py:
```diff
@@ -0,0 +1,249 @@
+    ]
+
+    @property
+    def _ccxt_config(self) -> dict:
+        config = {}
+        if self.trading_mode == TradingMode.SPOT:
+            config.update({"options": {"defaultType": "spot"}})
+        elif self.trading_mode == TradingMode.FUTURES:
+            config.update({"options": {"defaultType": "swap"}})
+        config.update(super()._ccxt_config)
+        return config
```
Is this really necessary?
this is the default, really.
Overriding for spot is only necessary if spot is not the default in ccxt - which i don't think is the case for bitget.
Automatically this means - you can remove this whole property ;)

yili1992 replied:
get it, i will remove it

Fat215 replied:
Hilo 1992

xmatthias commented on freqtrade/exchange/bitget.py:
```diff
@@ -0,0 +1,249 @@
+                    logger.info("Bitget: Position mode set to one-way.")
+
+                    # Ensure we're using the futures API
+                    self._api.options["defaultType"] = "swap"
+                else:
+                    # Ensure we're using the spot API
+                    self._api.options["defaultType"] = "spot"
```
definitely not necessary - and borderline wrong the way it's done here...

xmatthias commented on freqtrade/exchange/exchange.py:
```diff
@@ -1686,10 +1686,14 @@ def fetch_positions(self, pair: str | None = None) -> list[CcxtPosition]:
             return []
         try:
+            positions: list[CcxtPosition]
             symbols = []
             if pair:
                 symbols.append(pair)
-            positions: list[CcxtPosition] = self._api.fetch_positions(symbols)
+            if symbols:
+                positions = self._api.fetch_positions(symbols)
+            else:
+                positions = self._api.fetch_positions()
             self._log_exchange_response("fetch_positions", positions)
```
Any particular reason to have this? 
This seems to be a pointless change - as ccxt does support passing in an empty list / none.
if that's not working, its a bug you should report to ccxt instead.

yili1992 replied:
Because i found that bitget to call fetch_positions has a problem if symbols = [],
ccxt/bitget.py
```pyhon
   def fetch_positions(self, symbols: Strings = None, params={}) -> List[Position]:  
	...
        if symbols is not None:
            first = self.safe_string(symbols, 0)
            sandboxMode = self.safe_bool(self.options, 'sandboxMode', False)
            if sandboxMode:
                sandboxSymbol = self.convert_symbol_for_sandbox(first)
                market = self.market(sandboxSymbol)
            else:
                market = self.market(first)`
	....
```
_symbols is not None_  widely used in other exchanges in ccxt
So I think _self._api.fetch_positions()_ might be better than _self._api.fetch_positions([])_.

xmatthias replied:
If ccxt has a problem with this on bitget, please open an issue with ccxt to have this properly fixed.

It's working for **all** other exchanges that we support futures with - i don't see why we should make an exception for bitget just because of a bug in ccxt - which wasn't even reported.
We can talk again if they refuse to fix it - but please take the proper route instead of implementing dumb workarounds in freqtrade.

yili1992 replied:
I've retested this function recently, and I found no issues. I'll revert the code and remove my changes.

xmatthias commented on freqtrade/exchange/bitget.py:
```diff
@@ -0,0 +1,249 @@
+
+        return orders
+
+    def fetch_order(self, order_id: str, pair: str, params: dict | None = None) -> dict:
+        order = super().fetch_order(order_id, pair, params)
+        if not hasattr(order, "status"):
+            return order
+        if order["status"] == "closed" and order.get("average") is None and order.get("price"):
+            order["average"] = order["price"]
+        return order
```
Does this exchange really need this? How did you test that?

xmatthias commented on freqtrade/exchange/bitget.py:
```diff
@@ -0,0 +1,249 @@
+        if self.trading_mode == TradingMode.SPOT:
+            return float("inf")  # Not actually inf, but this probably won't matter for SPOT
+
+        if pair not in self._leverage_tiers:
+            return float("inf")
```
Definitely no.
If the exchange supports leverage tiers and the pair is not in leverage tiers, then the default is certainly not inf - it's 1.

I don't really see why this will need to be overwritten though?

yili1992 replied:
I referred to implement of OKX for this function.

xmatthias commented on freqtrade/exchange/bitget.py:
```diff
@@ -0,0 +1,249 @@
+        orders = []
+
+        while since < dt_now():
+            until = since + timedelta(days=7, minutes=-1)
+            orders += super().fetch_orders(pair, since, params={"until": dt_ts(until)})
+            since = until
+
```
Does bitget only have a max range of 7 days?

yili1992 replied:
I went back to review the document. I don't need to rewrite fetch_order and fetch_orders. i will remove it

xmatthias commented on freqtrade/exchange/bitget.py:
```diff
@@ -0,0 +1,249 @@
+
+            pair, _, candle_type, data, partial_candle = pair_data
+
+            if candle_type not in (CandleType.FUNDING_RATE):
```
I'm ... not entirely sure i understand why you think `_async_get_historic_ohlcv` must be re-implemented here.

Why ain't the regular implementation enough? is bitget that special when it comes to candles?

Fat215 replied:
Yes

yili1992 replied:
After syncing with the develop branch, I retest it, and I found no issues. I will remove these changes.

xmatthias commented on freqtrade/exchange/bitget.py:
```diff
@@ -0,0 +1,249 @@
+            )
+            raise
+
+    async def _estimate_current_candle(
```
fetch_ohlcv returns candles.
why do we need to aggregate these again ?

Fat215 replied:
Yes

yili1992 replied:
I discovered through the testcase that I couldn't obtain the latest candle, so I overwrote the function to calculate the latest candle myself.

xmatthias replied:
what's the "latest" candle for you in this case?

If it's about the open candle - then there's a setting you can switch to indicate that the exchange doesn't provide the "unfinished" candle (don't remember the name, but it's off for every supported exchange).

i'm not entirely sure this makes sense still. 
fetch_ohlcv from ccxt will provide the data in the correct format (and timerange) - if it doesn't - it's a ccxt bug, not something we work around here.

yili1992 replied:
I have submitted a bug to CCXT. https://github.com/ccxt/ccxt/issues/25289

xmatthias replied:
well I don't think ccxt is going to be of much help in this case (yes, it took some looking at this myself).

if you do the call for index (for example) manually - you get https://api.bitget.com/api/v2/mix/market/history-index-candles?symbol=BTCUSDT&granularity=1H&limit=100&productType=USDT-FUTURES

If you now look at the very last timestamp - then you'll see that it'll have this last timestamp missing.
As such - it's an exchange limitation - you'll have to talk to the exchange support for help.

That doesn't mean i'm willing to support or maintain such code here though.

mark / index candles ain't "really" used for anything other than funding fee / liquidation price calculation. 
If this is the workaround we must have - then we'll not be supporting bitget futures.

(aside from the willingness to support this - it's wrong and will mash mark / index candles together with regular/normal candles - making this an even more "no-go" approach).

xmatthias replied:
Looking at your CCXT issue - it doesn't seem to be an issue though - freqtrade wouldn't use an unfinished candle anyway (at 16:25 - the 15:00 candle is the latest closed candle) ... so i'm honestly not sure what you're trying to accomplish with that odd workaround here.

yili1992 replied:
Because this issue would cause the check in **test_ccxt_fetch_mark_price_history** to fail
` assert mark_candles[mark_candles["date"] == this_hour].iloc[0]["open"] != 0.0`
Currently, other exchanges like Binance are successful in this test. If Freqtrade wouldn't use an unfinished candle, could I modify this unit test to accommodate Bitget?

xmatthias replied:
no - this test tests a very specific thing - if it fails - then something is wrong.

I think you should instead work on understanding the actual problem (i don't think you did) - rather than trying random fixes.

For my understanding - is **this** test the reason for this whole method??

xmatthias replied:
@yili1992 i don't really see this point as resolved - and i don't like to include this method as i don't understand why it's necessary (i hence never looked at the implementation details - but i know it's absolutely wrong).

xmatthias reviewed (changes requested):
Ad hock - there seems a lot of unnecessary code (or at least code that needs explanation as to why it's there).

As is - it's not something we'll be accepting - but i've posted a few questions around the code - which either require changes (for things that are unnecessary) - or where it's unclear if it's just a copy/paste error.

Commit:
feat: fix comments

xmatthias commented on freqtrade/exchange/bitget.py:
```diff
@@ -50,41 +49,6 @@ class Bitget(Exchange):
-            if not self._config["dry_run"]:
-                if self.trading_mode == TradingMode.FUTURES:
-                    # Set position mode to one-way (hedged = False)
-                    # self._api.set_position_mode(False, None, {'productType': 'USDT-FUTURES'})
```
I didn't necessarily mean this line - just the "defaultType" settings below.
Checking the position mode  _might_ be fine - though in reality, you'll probably want to query the position mode (check if it's set to one-way and fail to start otherwise).

Force-setting it can fail miserably if there's an open position already - either failing to switch it - or by force-closing all open positions (at least that's my experience with other exchanges).

Commit:
chore: clean code

jerome-benoit commented:
@yili1992: did you have time to address the review comments since? I'm willing to either take over or help at integrating that exchange in freqtrade. 
For a lot of EU residents, all the CEX officially supported does not allow to trade futures in some countries. Bitget is one of the few still permitting it everywhere (at least for now). hyperliquid DEX allows it but the lack of candlesticks history make some use cases (freqai) requiring a complex setup, even for just backtesting.

Fat215 reviewed (commented):
Import Binance

Fat215 commented:
Add bitget

Fat215 commented:
Hilo 1992

yili1992 commented:
> @yili1992: did you have time to address the review comments since? I'm willing to either take over or help at integrating that exchange in freqtrade. For a lot of EU residents, all the CEX officially supported does not allow to trade futures in some countries. Bitget is one of the few still permitting it everywhere (at least for now). hyperliquid DEX allows it but the lack of candlesticks history make some use cases (freqai) requiring a complex setup, even for just backtesting.

@jerome-benoit ,  I'm late for reviewing comments. I would be pleased if you could take over or help with integrating that exchange.

jerome-benoit commented:
> @jerome-benoit , I'm late for reviewing comments. I would be pleased if you could take over or help with integrating that exchange.

PRs on top of your current branch will be fine?

yili1992 commented:
> > @jerome-benoit , I'm late for reviewing comments. I would be pleased if you could take over or help with integrating that exchange.
> 
> PRs on top of your current branch will be fine?

great, it's ok

jerome-benoit committed:
Merge branch 'freqtrade:develop' into develop

xmatthias commented on freqtrade/exchange/bitget.py:
```diff
@@ -0,0 +1,221 @@
+    ]
+
+    @property
+    def _ccxt_config(self) -> dict:
+        config = {}
+        if self.trading_mode == TradingMode.FUTURES:
+            config.update({"options": {"defaultType": "swap"}})
+        config.update(super()._ccxt_config)
+        return config
```
Well that's the default from upstream - so this whole property ain't necessary in this subclass.

yili1992 replied:
removed

4f1sh3r commented:
> @yili1992: did you have time to address the review comments since? I'm willing to either take over or help at integrating that exchange in freqtrade. For a lot of EU residents, all the CEX officially supported does not allow to trade futures in some countries. Bitget is one of the few still permitting it everywhere (at least for now). hyperliquid DEX allows it but the lack of candlesticks history make some use cases (freqai) requiring a complex setup, even for just backtesting.

@jerome-benoit: I would absolutely love that! I am lacking skills to help implementing Bitget futures trading, but I could help with testing! Thanks in advance for your efforts!

jerome-benoit commented:
> @jerome-benoit: I would absolutely love that! I am lacking skills to help implementing Bitget futures trading, but I could help with testing! Thanks in advance for your efforts!

I'm a very busy lately but I've not forgotten. And I'm not very inclined to contribute some code that project right now.

Bluestart83 commented:
I've tested it successfully!
Work far better than buggy binance spot that is unable to close an order due to lack of money! In Europe futures are banned of binance, so Bitget the second one in liquidity after Binance is a must have!

xmatthias commented:
Well unless the review comments are adressed - this ain't something to look at again, really.
And without activity - we'll be closing this PR as it's going to confuse people about availability of this exchange.

JarosLucky committed:
Not necessary overload

That's the default from upstream - so this whole property ain't necessary in this subclass

Commit:
feat: clean code for review comments

Commit:
Merge branch 'develop' of github.com:freqtrade/freqtrade into develop

yili1992 committed:
Merge pull request #2 from JarosLucky/patch-1

Not necessary overload

Commit:
Merge remote-tracking branch 'origin/develop' into develop

Commit:
chore: remove not necessary overload

yili1992 commented:
> Well unless the review comments are adressed - this ain't something to look at again, really. And without activity - we'll be closing this PR as it's going to confuse people about availability of this exchange.

@xmatthias I was too busy for a while before. Recently, I synchronized the branch and modified my code according to your comments. I also reran the unit tests. To my delight, those override functions were indeed unnecessary. Thank you again for your review.

Commit:
test: fix ut

Commit:
feat: override get history ohlcv for bitget mark & index

xmatthias commented on tests/exchange_online/test_ccxt_compat.py:
```diff
@@ -111,7 +111,11 @@ def test_ccxt_my_trades_parse(self, exchange: EXCHANGE_FIXTURE_TYPE):
             pair = "SOL/USDT"
             for trade in trades:
-                po = exch._api.parse_trade(trade)
+                if exchange_name == "bitget":
+                    market = exch._api.markets[pair]
+                    po = exch._api.parse_trade(trade, market)
+                else:
+                    po = exch._api.parse_trade(trade)
```
there shouldn't be an exception for this.
the ccxt default is `undefined` (or None in python) - so my assumption is that it'll work without this, too.
If that's not the case - then all other exchanges should have no problem WITH the market provided, either (that'd be the change i'd do here).

If either don't work (as in - bitget requires markets, other exchanges fail if market is provided) - then it's a ccxt bug.

xmatthias commented on freqtrade/exchange/bitget.py:
```diff
@@ -0,0 +1,213 @@
+    """
+
+    _ft_has: FtHas = {
+        "ohlcv_candle_limit": 1000,
```
You can omit ohlcv_candle_limit - it's taken from the ccxt "describe" - which also has it at 1000.
Also,  ohlcv_has_history defaults to true anyway ... so it's not necessary (especially not in the futures dict - which is added" on top" of the regular ftHas structure.

yili1992 replied:
got it

xmatthias committed:
chore: transfer "funding_fee_candle_limit" to ft_has_futures structure

xmatthias committed:
chore: cleanup some props that are default anyway

xmatthias commented on freqtrade/exchange/bitget.py:
```diff
@@ -0,0 +1,213 @@
+        # For other candle types, use the default or previously defined limit
+        return super().ohlcv_candle_limit(timeframe, candle_type, since_ms)
+
+    def get_funding_fees(
+        self, pair: str, amount: float, is_short: bool, open_date: datetime
+    ) -> float:
+        if self.trading_mode == TradingMode.FUTURES:
+            try:
+                return self._fetch_and_calculate_funding_fees(pair, amount, is_short, open_date)
+            except ExchangeError:
+                logger.warning(f"Could not update funding fees for {pair}.")
+        return 0.0
```
I'm not sure i understand the reason for this.

it seems to copy the logic from bybit - which requires emulation all the time - but ccxt's bitget does have `fetch_funding_history` - so the parent method should work fine?

yili1992 replied:
I will retest it

xmatthias commented on freqtrade/exchange/bitget.py:
```diff
@@ -0,0 +1,213 @@
+                logger.warning(f"Could not update funding fees for {pair}.")
+        return 0.0
+
+    def get_max_pair_stake_amount(self, pair: str, price: float, leverage: float = 1.0) -> float:
+        if self.trading_mode == TradingMode.SPOT:
+            return float("inf")  # Not actually inf, but this probably won't matter for SPOT
+
+        if pair not in self._leverage_tiers:
+            return 1.0
+
+        pair_tiers = self._leverage_tiers[pair]
+        return pair_tiers[-1]["maxNotional"] / leverage
```
I'm ... not convinced this is correct (the whole method).

It's disabling spot limits completely  - though the exchange provides them just fine (which is definitely wrong) - so "parent" should work perfectly fine.
Also for non-spot - i'm not sure i understand what makes this different to the parent method. It's simpler, and ignores a lot more stuff - but that doesn't make it correct.

xmatthias reviewed (changes requested):
I made a few cleanups - though there's still some way to go (it's still implementing way too much without real explanation as to why).

the exchange subclass should be as minimal as possible to work - this seems to implement random stuff (e.g. `get_max_pair_stake_amount` - which seems to just be there to skip correct generic logic in the parent class without reason)

xmatthias committed:
Merge branch 'develop' into yili1992/develop

Commit:
test: adjust unit test

xmatthias commented on tests/exchange/test_bitget.py:
```diff
@@ -360,20 +353,20 @@ def test_bitget_ohlcv_candle_limit(default_conf, mocker):
     for timeframe in timeframes:
-        assert exchange.ohlcv_candle_limit(timeframe, CandleType.SPOT) == 1000
-        assert exchange.ohlcv_candle_limit(timeframe, CandleType.FUTURES) == 1000
+        assert exchange.ohlcv_candle_limit(timeframe, CandleType.SPOT) == 500
```
This is incorrect - bitget yields 1000 candles.
If your (local!) environment yields 500 candles, you'll have to update ccxt to match the verison of this branch - as your ccxt version then doesn't have the "features" object yet - which suggests it's a pretty old version, too.

xmatthias commented on freqtrade/exchange/bitget.py:
```diff
@@ -27,11 +27,11 @@ class Bitget(Exchange):
     _ft_has: FtHas = {
         "order_time_in_force": ["GTC", "FOK", "IOC"],
         "trades_has_history": True,
+        "funding_fee_candle_limit": 100,
```
This should stay in the "futures" ft_has.
funding fees are irrelevant for spot trading - so this doesn't need to be set in the spot settings, either.

If this caused a problem in tests (i suspect that's why you moved it) - then your test setup is wrong.

Commit:
feat: change fetch ohlcv limit

xmatthias commented on freqtrade/exchange/exchange.py:
```diff
@@ -474,7 +474,7 @@ def ohlcv_candle_limit(
         fallback_val = self._ft_has.get("ohlcv_candle_limit", ccxt_val)
         if candle_type == CandleType.FUNDING_RATE:
-            fallback_val = self._ft_has.get("funding_fee_candle_limit", fallback_val)
+            fallback_val = self._ft_has_futures.get("funding_fee_candle_limit", fallback_val)
```
This is an absolute wrong change - please revert.

I understand **why** you did it - but it shows that your test setup is wrong (as pointed out previously).
for a futures exchange - the content of `_ft_has_futures` is moved to `_ft_has`.
The requirement for this change shows that your test does test futures things - but doesn't initialize the exchange object as futures object.

your fix for this **must** be in the test - not anywhere else.

xmatthias reviewed (changes requested):
Unfortunately still a few unacceptable changes. 
I tried to explain the problem you tried to wrongly fix as best as possible - but if you need help with it let me know (the fix for your "wrong funding length" problem needs to be in the test, not anywhere else).

Also - i still don't understand what `_estimate_current_candle` tries to accomplish. Is it really only there to adress the test failure in `test_ccxt_fetch_mark_price_history`, or is there other reasons? (i can take a look if that's the case - but i need to understand if that's the actual reason first).

yili1992 committed:
Update exchange.py

fix: revert to _ft_has

yili1992 committed:
Update test_bitget.py

test: fix unittest

yili1992 committed:
Merge pull request #3 from yili1992/yili1992-patch-1

Update exchange.py
